lib/Stan_math_code_gen.ml

```diff
@@ -116,61 +136,115 @@ let pp_arg ppf ((adtype, name, st), idx) =
   in
   pf ppf "const %a& %s" (pp_unsizedtype adstr) st name
 
-let pp_template_decls ppf =
-  pf ppf "@[<hov>template <%a>@]@ " (list ~sep:comma (fmt "typename %s"))
-
 let with_idx lst = List.(zip_exn lst (range 0 (length lst)))
 
 let%expect_test "with idx" =
   print_s [%sexp (with_idx (List.range 10 15) : (int * int) list)] ;
   [%expect {| ((10 0) (11 1) (12 2) (13 3) (14 4)) |}]
 
-let pp_error_wrapper ppf (pp_err, err_arg, pp_contents, contents_arg) =
-  string ppf "try " ;
-  pp_block ppf (pp_contents, contents_arg) ;
-  string ppf " catch const std::exception& e) " ;
-  pp_block ppf (pp_err, err_arg)
+let pp_decl ppf (vident, st) =
+  pf ppf "%a %s;" pp_prim_stantype (Ast.remove_size st) vident
+
+let with_no_loc stmt = {stmt; sloc= ""}
 
 let pp_located_msg ppf msg =
   pf ppf
     {|stan::lang::rethrow_located(
-    std::runtime_error(std::string(%s) + e.what(), current_statement__);
-// Next line prevents compiler griping about no return
-throw std::runtime_error("*** IF YOU SEE THIS, PLEASE REPORT A BUG ***");
-|}
+          std::runtime_error(std::string(%s) + e.what(), current_statement__));
+      // Next line prevents compiler griping about no return
+      throw std::runtime_error("*** IF YOU SEE THIS, PLEASE REPORT A BUG ***"); |}
   @@ Option.value ~default:"e" msg
 
-let pp_located_error ppf (pp_contents, contents_arg, msg) =
-  pp_error_wrapper ppf (pp_located_msg, msg, pp_contents, contents_arg)
+let maybe_templated_arg_types (args : formal_params) =
+  let is_autodiff (adtype, _, _) =
+    match adtype with Ast.AutoDiffable -> true | _ -> false
+  in
+  List.filter ~f:is_autodiff args |> List.mapi ~f:(fun i _ -> sprintf "T%d__" i)
```

Comment:
Getting these template parameters to match (and not clash with each other!) everywhere seems like one of the hard tasks in code gen for me. Is it obvious to you that your current strategy will work? Are you using the same naming scheme as in the existing code gen? (Just asking because I cannot assess the correctness of this myself right now so wanted to at least flag it as a potentially subtle point.)

Reply:
I don't have a global strategy on that. The trickiest seeming ones are to make sure each model method uses the same template parameters, and the easiest thing there is to just always use `T__` for the autodiff type. I think only within FunDefs do we need to worry, and here we probably need to note the interaction between VarDecl's being subtly incorrect (i.e. always being autodiffable) and FunDef's translation creating a variable number of template parameters and then not passing them into pp_statement anywhere.